In [1]:
import numpy as np

# Implementare operatie de flattening

In [2]:
arr = np.random.randn(4,4,64)

In [3]:
arr.shape

(4, 4, 64)

In [4]:
arr.flatten().shape

(1024,)

# Implementare CNN deep in Tensorflow

In [47]:
import tensorflow as tf
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense, Flatten, Conv2D, MaxPool2D, Input, Dropout
from tensorflow.keras.datasets.mnist import load_data

In [48]:
(X_train, y_train), (X_test, y_test) = load_data()

In [49]:
X_train = np.expand_dims(X_train, axis=-1)
X_test = np.expand_dims(X_test, axis=-1)

In [50]:
model = Sequential([
    Input(shape=(28, 28, 1)),
    Conv2D(32, (5,5), padding='valid', activation='relu'),
    MaxPool2D(pool_size=(2,2)),
    Conv2D(64, (5,5), padding='valid', activation='relu'),
    MaxPool2D(pool_size=(2,2)),
    Flatten(),
    Dense(256, activation='relu'),
    Dropout(0.2),
    Dense(10, activation='softmax')
])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

In [51]:
model.summary()

Model: "sequential_7"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_14 (Conv2D)           (None, 24, 24, 32)        832       
_________________________________________________________________
max_pooling2d_14 (MaxPooling (None, 12, 12, 32)        0         
_________________________________________________________________
conv2d_15 (Conv2D)           (None, 8, 8, 64)          51264     
_________________________________________________________________
max_pooling2d_15 (MaxPooling (None, 4, 4, 64)          0         
_________________________________________________________________
flatten_7 (Flatten)          (None, 1024)              0         
_________________________________________________________________
dense_14 (Dense)             (None, 256)               262400    
_________________________________________________________________
dropout_7 (Dropout)          (None, 256)              

In [52]:
model.fit(
    epochs=10,
    x=X_train,
    y=y_train,
    batch_size=128,
    validation_data=(X_test, y_test)
)

Epoch 1/10
469/469 [==============================] - 38s 79ms/step - loss: 2.2592 - accuracy: 0.8158 - val_loss: 0.0603 - val_accuracy: 0.9805
Epoch 2/10
469/469 [==============================] - 32s 68ms/step - loss: 0.0613 - accuracy: 0.9813 - val_loss: 0.0557 - val_accuracy: 0.9837
Epoch 3/10
469/469 [==============================] - 31s 66ms/step - loss: 0.0428 - accuracy: 0.9863 - val_loss: 0.0353 - val_accuracy: 0.9891
Epoch 4/10
469/469 [==============================] - 32s 68ms/step - loss: 0.0286 - accuracy: 0.9908 - val_loss: 0.0417 - val_accuracy: 0.9874
Epoch 5/10
469/469 [==============================] - 33s 69ms/step - loss: 0.0254 - accuracy: 0.9922 - val_loss: 0.0482 - val_accuracy: 0.9867
Epoch 6/10
469/469 [==============================] - 33s 71ms/step - loss: 0.0255 - accuracy: 0.9917 - val_loss: 0.0399 - val_accuracy: 0.9892
Epoch 7/10
469/469 [==============================] - 33s 71ms/step - loss: 0.0214 - accuracy: 0.9931 - val_loss: 0.0515 - val_accuracy:

In [53]:
preds = model.evaluate(X_test, y_test)

313/313 [==============================] - 2s 5ms/step - loss: 0.0509 - accuracy: 0.9884


# Tema optionala

Incercati sa modificati arhitectura de sus in asa fel incat acuratetea finala sa fie mai mare de 98.84%. :)